src/Lecture3.hs

```diff
@@ -52,201 +48,101 @@ data Weekday
 
 instance Monoid Reward where
+    mempty = Reward (Gold 0) False
 
-
-{- | 'List1' is a list that contains at least one element.
--}
 data List1 a = List1 a [a]
     deriving (Show, Eq)
 
--- | This should be list append.
-instance Semigroup (List1 a) where
+instance Semigroup(List1 a) where
+    (List1 l1 a1)<> (List1 l2 a2)= List1 l1 (a1 ++[l2] ++ a2)
 
 
 {- | Does 'List1' have the 'Monoid' instance? If no then why?
 
 instance Monoid (List1 a) where
 -}
+--Listl cannot have a Monoid instance since it cannot have zero elements
 
-{- | When fighting a monster, you can either receive some treasure or
-don't.
--}
 data Treasure a
     = NoTreasure
     | SomeTreasure a
     deriving (Show, Eq)
 
-{- | When you append multiple treasures for fighting multiple
-monsters, you should get a combined treasure and not just the first
-(or last one).
-
-🕯 HINT: You may need to add additional constraints to this instance
-  declaration.
--}
-instance Semigroup (Treasure a) where
-
-
-instance Monoid (Treasure a) where
-
-
-{- | Abstractions are less helpful if we can't write functions that
-use them!
-
-Implement a polymorphic function that takes three elements and appends
-together only different elements.
-
->>> appendDiff3 [1] [3, 2] [0, 5]
-[1,3,2,0,5]
->>> appendDiff3 [4] [2, 2] [2, 2]
-[4,2,2]
->>> appendDiff3 [1 .. 5] [1 .. 5] [1 .. 5]
-[1,2,3,4,5]
->>> appendDiff3 (Product 2) (Product 3) (Product 3)
-Product {getProduct = 6}
-
--}
-appendDiff3 = error "TODO"
-
-{-
+instance (Semigroup a) => Semigroup (Treasure a) where
+    NoTreasure <> SomeTreasure a = SomeTreasure a
+    NoTreasure <> NoTreasure = NoTreasure
+    SomeTreasure a <> NoTreasure = SomeTreasure a
+    SomeTreasure a1 <> SomeTreasure a2 = SomeTreasure (a1 <> a2)
 
-In the next block of tasks, implement 'Foldable' instances for all
-types that can have such an instance.
+instance (Semigroup a) => Monoid (Treasure a) where
+    mempty = NoTreasure
 
-♫ NOTE: Implement both 'foldr' and 'foldMap' methods. On the one hand,
-  'Foldable' is a big typeclass but lets focus on its small part to get
-  the main idea. On the other hand, these two methods are quite
-  different so it's a good practice.
-
-🕯 HINT: Check kinds of types to see whether it's possible to implement
-  an instance of 'Foldable'.
-
-🕯 HINT: If you don't feel comfortable with kinds yet, alternatively
-  you can try uncommenting each instance one by one and see the GHC
-  error. The compiler will "kindly" tell you if it's impossible to have
-  such an instance.
-
-🕯 HINT: Write explicit type signature of methods using InstanceSigs
-  (already enabled in this module).
-
-♫ NOTE: Since the instances are commented, the tests are also commented.
-  To run tests for your instances, go to the "test/Test/Lecture3.hs" file
-  and uncomment all commented tests. But do this only after you
-  implement instances! No spoilers :)
--}
+appendDiff3:: Eq a => Semigroup a => Monoid a => a -> a-> a-> a
```

Comment:
That's interesting syntax for writing several constraints 🤔 
Usually, multiple constraints are specified via `()` and separated by commas but the usage of multiple arrows is also allowed 👍🏻 

If you mention `Monoid`, you don't need to write `Semigroup` because `Monoid` implies `Semigroup`. However, in this situation it's better to remove the `Monoid` constraint instead and don't use `mempty` and rely only on `Semigroup` methods. If you have the `Monoid` constraint, function `appendDiff3` won't work with type like `List1` that don't have `Monoid` instance but it's possible to implement to allow this 🙂